Add `Nested` Documentation for 2.7 Related Features

forestmvey opened the pull request:
### Description
Added documentation for architectural changes from [PR-1490](https://github.com/opensearch-project/sql/pull/1490). The added features covers the `nested` function used in the `SELECT` clause of an SQL statement. Future work has placeholders in documentation and can be added to with proposals as development continues. 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

GumpacG reviewed (approved):
Looks great! Not sure if it's appropriate in these docs but maybe add a simple example of mapping, data, and expected result of a simple query for devs in the future who don't have knowledge of nested in OpenSearch. May be appropriate in the use case section.

forestmvey commented:
> Looks great! Not sure if it's appropriate in these docs but maybe add a simple example of mapping, data, and expected result of a simple query for devs in the future who don't have knowledge of nested in OpenSearch. May be appropriate in the use case section.

Good idea!

Yury-Fridlyand commented:
1. List new docs in `index.md`
2. Add links to demo you did for `nested`

matthewryanwells commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,413 @@
+    LogicalPlan --> OptimizedLogicalPlan: Optimize
+    OptimizedLogicalPlan --> PhysicalPlan:push down
+    note right of PhysicalPlan
+    NestedOperator stays in PhysicalPlan\nafter push down for post-processing.
```
Is the \n supposed to be there? It appears to be a newline character but considering I don't see any other \n throughout the rest of the document I think it might be accidental.

matthewryanwells commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,413 @@
+    NestedOperator stays in PhysicalPlan\nafter push down for post-processing.
+    end note
+
+    state "Logical Plan" as LogicalPlan:::blah
```
is `blah` something that we want to be in documentation? Something about it just seems like it shouldn't appear in official documentation, although I don't know what other thing would be better to replace it

matthewryanwells commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,413 @@
+        logState3 --> Relation
+    }
+
+    state "Optimized Logical Plan" as OptimizedLogicalPlan:::blah
```
Same comment as above

matthewryanwells commented on docs/dev/sql-nested-function-select-clause.md:
````diff
@@ -0,0 +1,413 @@
+Nested function class diagram for additional classes required for query execution. The `NestedAnalyzer` is a visitor for nested functions used in the SELECT clause to fulfill the `LogicalNested` LogicalPlan. After push down is successful the `NestedOperator` PhysicalPlan is used for object flattening of the OpenSearch response.
+
+```mermaid
+---
+title: Nested Select Clause Class Diagram
+---
````
On my machine it appears that the inclusion of this title section causes the graph to have a syntax error, removing these lines causes the graph to load properly.

<img width="665" alt="image" src="https://user-images.githubusercontent.com/121696727/236346922-73cb3ddd-8de1-4d6f-bb52-81f359edf5b5.png">

forestmvey replied:
If you go to the file [here](https://github.com/Bit-Quill/opensearch-project-sql/blob/7b302aff85cca44351bd2c0976c9d6582e9be525/docs/dev/sql-nested-function-select-clause.md#24-select-clause-nested-query-class-diagram), you can see that github renders the diagram correctly. I think you may need a newer version of mermaid. Using confluence which also renders this correctly I see we have version 9.4.3.

forestmvey commented:
> 1. List new docs in `index.md`
> 2. Add links to demo you did for `nested`

Thanks, I have added the files I added to `index.md`. There is already my demo video embedded in `sql-nested-function-select-clause.md`

acarbonetto commented on docs/dev/sql-nested-function-select-clause.md:
````diff
@@ -0,0 +1,476 @@
+Nested function state diagram illustrating states in SQL plugin for push down execution. The nested operator stays in the `PhysicalPlan` after push down for flattening operation in post-processing. See section [2.3](#24-select-clause-nested-query-class-diagram) for flattening sequence and description.
+
+```mermaid
+stateDiagram-v2
````
Can you update the classes in this diagram to refer to the actual names of the classes?  I think it's `OpenSearchIndexScanBuilder` for example.

acarbonetto commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,476 @@
+    LogicalPlan --> OptimizedLogicalPlan: Optimize
+    OptimizedLogicalPlan --> PhysicalPlan:push down
+    note right of PhysicalPlan
+      Note: NestedOperator stays in PhysicalPlan\nafter push down for post-processing.
```
I'm not a big can of this `note`.  Would it be better to include a state for `Request` where the ExecutionEngine opens the request, and then we have another state afterwards for post-request processing?  It might be nice to include the inner_hits flattening there (as part of the `ProjectOperator`).

forestmvey replied:
No point in repeating things in multiple diagrams. Would like them to stay basic and only illustrate things that are necessary or unique for this query. I have removed the note and it is noted in the description that the `NestedOperator` stays in the `Physical Plan Tree` for post-processing, and a link to the sequence diagram that illustrates the response flattening.

acarbonetto commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,476 @@
+### 2.3 Sequence Diagram for Nested SELECT Clause Post-processing
+Nested function sequence diagram illustrating the flattening of the OpenSearch response. Flattening the response from OpenSearch changes the nested types structure by making the full path of an object the key, and the object it refers to the value.
+
+**Sample input:**
```
Could you include an example where we do a little more flattening?  Such as two nested objects with two fields each to show how we do the cartesian product?

acarbonetto commented on docs/dev/sql-nested-function-select-clause.md:
````diff
@@ -0,0 +1,476 @@
+ProjectOperator-->>-OpenSearchExecutionEngine:ExprValue
+```
+
+#### 2.4 Select Clause Nested Query Class Diagram
````
```suggestion
### 2.4 Select Clause Nested Query Class Diagram
```

acarbonetto reviewed (approved):
Looks good.  Please address comments.

forestmvey commented on docs/dev/sql-nested-function.md:
```diff
@@ -0,0 +1,91 @@
+## Description
+
+The nested function in SQL and PPL maps to the nested query DSL in the OpenSearch query engine. A nested query is used to search nested object field types in an index. If an object matches the search, the nested query returns the root parent document. Nested inner objects are returned as inner hits in the query result.
```
I would like to modularize as much as I can and handle the syntax and specific clause/query language implementations in each of their own designated documents. This document isn't specific to the `SELECT` clause which does specify the syntax.

forestmvey commented on docs/dev/sql-nested-function.md:
```diff
@@ -0,0 +1,91 @@
+### 1.1 Problem Statement
+
+**1. The V2 engine lacks legacy functionality for the nested function** - 
+The `nested` function is not present in the V2 engine and must be brought forward from the legacy engine to support user queries for nested object field types in SQL.
```
Revised to be more general about supporting OS nested object type.

MaxKsyunz commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,484 @@
+                                        }
+                                    },
+                                    "path": "message",
+                                    "ignore_unmapped": false,
+                                    "score_mode": "none",
+                                    "boost": 1.0,
+                                    "inner_hits": {
+                                        "ignore_unmapped": false,
+                                        "from": 0,
+                                        "size": 3,
+                                        "version": false,
+                                        "seq_no_primary_term": false,
+                                        "explain": false,
+                                        "track_scores": false,
+                                        "_source": {
+                                            "includes": [
+                                                "message.info"
+                                            ],
+                                            "excludes": []
+                                        }
+                                    }
+                                }
+                            }
+                        ],
+                        "adjust_pure_negative": true,
+                        "boost": 1.0
+                    }
+                }
+            ],
+            "adjust_pure_negative": true,
+            "boost": 1.0
+        }
+    },
+    "_source": {
+        "includes": [
+            "message.info"
+        ],
+        "excludes": []
+    }
+}
```
There's a lot going on in this DSL query and it's hard to understand which parts were added due to nested expression.

Is it possible to delete any of the properties? For example,  by default `adjust_pure_negative` is set to `true` by OpenSearch. I expect the other boolean properties are being set to their default values as well. Even if explain endpoint shows these values, as a reader it will help if they are not there.

MaxKsyunz commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,484 @@
+---
+classDiagram
+
+    NestedAnalyzer~AbstractNodeVisitor~-->LogicalNested~LogicalPlan~
+    LogicalNested-->NestedOperator~PhysicalPlan~
```
What do you want to express with this graph?

`NestedAnalyzer` creates a `LogicalNested`, so  that association makes sense, but there is no static link between `LogicalNested` and `NestedOperator`.

MaxKsyunz replied:
Maybe include the class that creates it?
IntelliJ creates something like this

```mermaid
classDiagram
direction BT
class DefaultImplementor {
  + visitNested(LogicalNested, C) PhysicalPlan

}
class LogicalNested {
  + getProjectList() List~NamedExpression~
  + accept(LogicalPlanNodeVisitor, C) R
  + addFields(Map~String_ReferenceExpression~) void
}
class NestedAnalyzer {
  + analyze(UnresolvedExpression, AnalysisContext) LogicalPlan
  + visitFunction(Function, AnalysisContext) LogicalPlan
  + visitAlias(Alias, AnalysisContext) LogicalPlan
}
class NestedOperator

DefaultImplementor~C~  ..>  NestedOperator : «create»
NestedAnalyzer  ..>  LogicalNested : «create»


```

MaxKsyunz commented on docs/dev/sql-nested-function-select-clause.md:
```diff
@@ -0,0 +1,484 @@
+
+### Release Schedule
+See Issues Tracked under [Issue 1111](https://github.com/opensearch-project/sql/issues/1111) for related PR's and information:
+- [x] **Phase 1:** Add support for nested function used in SELECT clause.
+- [ ] **Phase 2:** Add support for `*` at ending level of field parameter in SELECT clause nested function.
```
I'd remove this -- referencing the PR is sufficient.
